File: scripts/image-generation/convert-to-webp.py
"""
Convert PNG beard sketches to WebP format
Usage: python convert-to-webp.py
"""

from PIL import Image, ImageOps
import os

RAW_DIR = './raw'
OUTPUT_DIR = '../../frontend/public/assets/sketches'

# File mapping (raw name -> output name)
FILE_MAP = {
    'Stubble (3-Day Shadow).png': 'stubble-3day.webp',
    'van-dyke.png': 'van-dyke.webp',
    'full-beard.png': 'full-beard.webp',
    'ducktail.png': 'ducktail.webp',
}

def convert_image(input_name, output_name):
    input_path = os.path.join(RAW_DIR, input_name)
    output_path = os.path.join(OUTPUT_DIR, output_name)

    if not os.path.exists(input_path):
        print(f"  [SKIP] {input_name} - file not found")
        return False

    try:
        # Open image
        img = Image.open(input_path)

        # Convert to grayscale
        img = img.convert('L')

        # Normalize (auto-contrast)
        img = ImageOps.autocontrast(img)

        # Resize to 800x800 with white background
        img.thumbnail((800, 800), Image.Resampling.LANCZOS)

        # Create white background and paste
        background = Image.new('L', (800, 800), 255)
        offset = ((800 - img.width) // 2, (800 - img.height) // 2)
        background.paste(img, offset)

        # Convert back to RGB for WebP
        rgb_img = Image.new('RGB', (800, 800), (255, 255, 255))
        rgb_img.paste(Image.merge('RGB', [background, background, background]))

        # Save as WebP
        rgb_img.save(output_path, 'WEBP', quality=85)

        file_size = os.path.getsize(output_path) / 1024
        print(f"  [OK] {input_name} -> {output_name} ({file_size:.0f}KB)")
        return True

    except Exception as e:
        print(f"  [ERR] Error converting {input_name}: {e}")
        return False

def main():
    print("\n[*] Converting PNG to WebP...\n")

    # Ensure output directory exists
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    success = 0
    failed = 0

    for input_name, output_name in FILE_MAP.items():
        if convert_image(input_name, output_name):
            success += 1
        else:
            failed += 1

    print(f"\n[OK] Done: {success} converted, {failed} skipped/failed\n")

if __name__ == '__main__':
    main()

File: scripts/image-generation/convert-to-webp.js
/**
 * Convert PNG/JPG beard sketches to WebP format
 * Scans all 16 style folders in ./raw/ automatically
 * Usage: node convert-to-webp.js
 */

const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const RAW_DIR = './raw';
const OUTPUT_DIR = '../../frontend/public/assets/sketches';
const THUMB_DIR = '../../frontend/public/assets/thumbnails';
const MAIN_SIZE = 800;
const THUMB_SIZE = 200;
const QUALITY = 85;

function discoverStyles() {
  return fs.readdirSync(RAW_DIR, { withFileTypes: true })
    .filter(d => d.isDirectory() && /^\d{2}-/.test(d.name))
    .sort((a, b) => a.name.localeCompare(b.name))
    .map(d => {
      const slug = d.name.replace(/^\d{2}-/, '');
      const folderPath = path.join(RAW_DIR, d.name);
      const images = fs.readdirSync(folderPath)
        .filter(f => /\.(png|jpg|jpeg)$/i.test(f))
        .sort();
      return { folder: d.name, slug, folderPath, images };
    });
}

async function convertImage(inputPath, outputPath, size) {
  await sharp(inputPath)
    .resize(size, size, { fit: 'contain', background: { r: 255, g: 255, b: 255, alpha: 1 } })
    .grayscale()
    .normalise()
    .webp({ quality: QUALITY })
    .toFile(outputPath);

  return fs.statSync(outputPath).size;
}

async function processStyle(style) {
  if (style.images.length === 0) {
    console.log(`  -- ${style.folder} (${style.slug}) — no images, skipping`);
    return null;
  }

  // Use first image found in the folder
  const srcFile = style.images[0];
  const inputPath = path.join(style.folderPath, srcFile);
  const mainOut = path.join(OUTPUT_DIR, `${style.slug}.webp`);
  const thumbOut = path.join(THUMB_DIR, `${style.slug}-thumb.webp`);

  try {
    const mainSize = await convertImage(inputPath, mainOut, MAIN_SIZE);
    const thumbSize = await convertImage(inputPath, thumbOut, THUMB_SIZE);

    console.log(
      `  OK ${style.folder} — ${srcFile} -> ${style.slug}.webp (${Math.round(mainSize / 1024)}KB) + thumb (${Math.round(thumbSize / 1024)}KB)`
    );

    return {
      slug: style.slug,
      main: `/assets/sketches/${style.slug}.webp`,
      thumbnail: `/assets/thumbnails/${style.slug}-thumb.webp`,
      exists: true,
      thumbnailExists: true,
    };
  } catch (err) {
    console.error(`  FAIL ${style.folder} — ${err.message}`);
    return {
      slug: style.slug,
      main: `/assets/sketches/${style.slug}.webp`,
      thumbnail: `/assets/thumbnails/${style.slug}-thumb.webp`,
      exists: false,
      thumbnailExists: false,
    };
  }
}

async function main() {
  console.log('\nConverting beard sketches to WebP...\n');

  // Ensure output directories exist
  for (const dir of [OUTPUT_DIR, THUMB_DIR]) {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  const styles = discoverStyles();
  console.log(`Found ${styles.length} style folders\n`);

  const manifest = [];
  let success = 0;
  let skipped = 0;

  for (const style of styles) {
    const result = await processStyle(style);
    if (result) {
      manifest.push(result);
      if (result.exists) success++;
      else skipped++;
    } else {
      skipped++;
      manifest.push({
        slug: style.slug,
        main: `/assets/sketches/${style.slug}.webp`,
        thumbnail: `/assets/thumbnails/${style.slug}-thumb.webp`,
        exists: false,
        thumbnailExists: false,
      });
    }
  }

  // Write manifest
  const manifestPath = path.join(OUTPUT_DIR, 'manifest.json');
  const manifestData = {
    styles: manifest,
    generated: new Date().toISOString(),
  };
  fs.writeFileSync(manifestPath, JSON.stringify(manifestData, null, 2));

  console.log(`\nDone: ${success} converted, ${skipped} skipped`);
  console.log(`Manifest written to ${manifestPath}\n`);
}

main().catch(console.error);
